### PR TITLE
PROV: Move bio_prov.c from libcommon.a to libfips.a / libnonfips.a

### PR DESCRIPTION
**Pull request by levitte**
libcommon.a is FIPS agnostic, while libfips.a and libnonfips.a are
FIPS / non-FIPS specific.  Since bio_prov.c checks FIPS_MODULE, it
belongs to the latter.

Along with this, a bit more instruction commentary is added to
providers/build.info.

### CODE REVIEW

**Comment by levitte** [comment]
@InfoHunter, if you could test this on your MacBook, that would be great

**Comment by InfoHunter** [comment]
> @InfoHunter, if you could test this on your MacBook, that would be great

Done testing, very smooth

**Comment by InfoHunter** [comment]
The travis failure seems not relevant.

File: providers/build.info [hunk 1]
```diff
@@ -27,6 +27,26 @@
 # libnonfips.a          Corresponds to libfips.a, but built with
 #                       FIPS_MODULE undefined.  The default and legacy
 #                       providers use this.
+#
+# This is how different provider modules should be linked:
+#
+# FIPS:
+#     -o fips.so {object files...} libimplementations.a libcommon.a libfips.a
+# Non-FIPS:
+#     -o fips.so {object files...} libimplementations.a libcommon.a libnonfips.a
```
Comment (richsalz) [inline providers/build.info:36]: Are you sure about fips.so here?

Comment (levitte) [reply providers/build.info:36]: Ah, no :laughing:

**Comment by slontis** [comment]
Why was this only an issue on Mac??

**Comment by levitte** [comment]
> Why was this only an issue on Mac??

I think Mac is more diligent than Linux regarding unresolved symbols in a module at link time.  Unfortunately, the `-z defs` we use on Linux is only good for object files used directly on the command line, i.e. doesn't apply to static libraries...

So, since `bio_prov.c` had a section of code wrapped with `#ifndef FIPS_MODULE`, and `libcommon.a` has object files compiled with that macro undefined, we ended up with a FIPS module with a few symbols unresolved, all of those referred to in that section of `bio_prov.c`.  Ignored on Linux, but cause for error on Mac.

**Comment by levitte** [comment]
@InfoHunter, I did a small correction requested by @richsalz...  a typo.  Does your approval stand?

**Comment by levitte** [comment]
Merged

904f42509f8d5e6210113e49a7e41ed2b1dd5a81 PROV: Move bio_prov.c from libcommon.a to libfips.a / libnonfips.a

**Pull request closed without merging**